File: paper_artifacts/experiments/llr9/artifacts/sources/oss120b-c-skills/wf_triangular__submitted.c
#include <stdint.h>
#include <omp.h>

void wf_triangular_fp64(double * restrict a, int64_t LEN_2D, uint8_t * restrict workspace, int64_t workspace_bytes) {
    // Triangular wavefront: a[i,j] depends on a[i-1,j] and a[i,j-1]
    // Parallelize over anti-diagonals (i + j constant)
    const int64_t N = LEN_2D;
    // The smallest sum of indices (i+j) is 2 (i=1,j=1). The largest is 2*(N-1).
    #pragma omp parallel
    {
        for (int64_t s = 2; s <= 2 * (N - 1); ++s) {
            // Determine i range for this anti-diagonal respecting the triangular domain (j >= i).
            int64_t i_start = s - (N - 1);
            if (i_start < 1) i_start = 1; // i must be at least 1
            int64_t i_end = s / 2; // floor division ensures i <= j
            if (i_end > N - 1) i_end = N - 1;
            // Parallel work over independent points on this anti-diagonal.
            #pragma omp for schedule(static)
            for (int64_t i = i_start; i <= i_end; ++i) {
                int64_t j = s - i; // column index
                // Update cell (i,j)
                a[i * N + j] = a[i * N + j] + a[(i - 1) * N + j] + a[i * N + (j - 1)];
            }
        }
    }
}

File: paper_artifacts/experiments/llr9/artifacts/sources/oss120b-fortran-skills/wf_triangular__submitted.f90
subroutine wf_triangular_fp64(a, n, workspace, workspace_size) bind(C)
  use iso_c_binding
  integer(c_int64_t), value, intent(in) :: n
  real(c_double), intent(inout) :: a(n, n)
  type(c_ptr), value, intent(in) :: workspace
  integer(c_int64_t), value, intent(in) :: workspace_size
  integer(c_int64_t) :: i, j, t, i_start, i_end
  ! Wavefront triangular update for j >= i using wavefront parallelism
  !$omp parallel private(i, j, t, i_start, i_end)
  do t = 4, 2*n
    i_start = t - n
    if (i_start < 2_c_int64_t) i_start = 2_c_int64_t
    i_end = t / 2
    if (i_end > n) i_end = n
    !$omp do schedule(static)
    do i = i_start, i_end
      j = t - i
      a(j, i) = a(j, i) + a(j, i-1) + a(j-1, i)
    end do
    !$omp end do
  end do
  !$omp end parallel
end subroutine wf_triangular_fp64
